ddtrace/contrib/jinja2/patch.py

```diff
@@ -0,0 +1,131 @@
+import jinja2
+from wrapt import wrap_function_wrapper as _w
+
+from ddtrace import config
+
+from ...ext import http
+from ...utils.formats import asbool, get_env
+from ...pin import Pin
+from ...utils.wrappers import unwrap as _u
+from .constants import DEFAULT_SERVICE, DEFAULT_TEMPLATE_NAME
+
+
+# default settings
+config._add('jinja2',{
+    'service_name': get_env('jinja2', 'service_name', DEFAULT_SERVICE),
+    'inherit_service': asbool(get_env('jinja2', 'inherit_service', True)),
+})
+
+
+def patch():
+    if getattr(jinja2, "__datadog_patch", False):
+        # already patched
+        return
+    setattr(jinja2, "__datadog_patch", True)
+    Pin(
+        service=config.jinja2['service_name'],
+        app='jinja2',
+        app_type=http.TEMPLATE,
+        _config=config.jinja2,
+    ).onto(jinja2.environment.Environment)
+    _w(jinja2, "environment.Template.render", _wrap_render)
+    _w(jinja2, "environment.Template.generate", _wrap_generate)
+    _w(jinja2, "environment.Environment.compile", _wrap_compile)
+    _w(jinja2, "environment.Environment._load_template", _wrap_load_template)
+
+
+def unpatch():
+    if not getattr(jinja2, "__datadog_patch", False):
+        return
+    setattr(jinja2, "__datadog_patch", False)
+    _u(jinja2.Template, "render")
+    _u(jinja2.Template, "generate")
+    _u(jinja2.Environment, "compile")
+    _u(jinja2.Environment, "_load_template")
+
+
+def _get_service_name(environment, span):
+    """if `inherit_service` is set then tries to get the service name of the
+    parent span.
+    The default value is set to `service_name` config variable (default: jinja2)
+    """
+    cfg = config.get_from(environment)
+    if cfg['inherit_service']:
+        if span._parent and span._parent.service:
+            return span._parent.service
+    return cfg['service_name']
+
+
+def _wrap_generate(wrapped, instance, args, kwargs):
+    pin = Pin.get_from(instance.environment)
+    if not pin or not pin.enabled():
+        for s in wrapped(*args, **kwargs):
+            yield s
+
+    template_name = instance.name or DEFAULT_TEMPLATE_NAME
+    with pin.tracer.trace("jinja2.generate", span_type=http.TEMPLATE) as span:
+        # update the span service name before doing any action
+        span.service = _get_service_name(instance.environment, span)
+        try:
+            for s in wrapped(*args, **kwargs):
```

Comment:
Same here, the function just needs to return a generator, which calling `wrapped(*args, **kwargs)` should do.